### PR TITLE
Add OpenAI component

### LINKED ISSUE #4567: Add OpenAI component
Today we have the [Azure.AI.OpenAI](https://github.com/dotnet/aspire/tree/main/src/Components/Aspire.Azure.AI.OpenAI) component when using Azure OpenAI.

Last week an official OpenAI library was introduced with https://devblogs.microsoft.com/dotnet/openai-dotnet-library/. We should consider having a plain `Aspire.OpenAI` component that wraps this library for the cases where an Aspire app wants to use OpenAI, but doesn't want to bring in Azure dependencies.

cc @sebastienros @tg-msft @davidfowl @mitchdenny

### PR DESCRIPTION
**Pull request by sebastienros**
Contributes to #4567 
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5621)

### CODE REVIEW

File: src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(NetCurrent)</TargetFramework>
```
Comment (radical) [inline src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj:4]: Use this property now which is set to `net8.0`, because `NetCurrent` is now `net9.0`.

```suggestion
    <TargetFramework>$(DefaultTargetFramework)</TargetFramework>
```

**Comment by eerhardt** [comment]
Is there also work to enable `Aspire.Azure.AI.OpenAI` to work with just an OpenAIClient when configured to talk to OpenAI or a local emulator like Ollama? Or do you see that as a separate PR?

File: src/Components/Telemetry.md [hunk 2]
```diff
@@ -257,6 +257,13 @@ Aspire.Npgsql.EntityFrameworkCore.PostgreSQL:
     - "db.client.connections.timeouts"
     - "db.client.connections.usage"
 
+Aspire.OpenAI:
+- Log categories:
```
Comment (eerhardt) [inline src/Components/Telemetry.md:261]:
```suggestion
- Log categories:
  - none
```

Comment (eerhardt) [reply src/Components/Telemetry.md:261]: FYI - @annelo-msft who is working on the ILogger support for System.ClientModel.

File: src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(DefaultTargetFramework)</TargetFramework>
+    <IsPackable>true</IsPackable>
+    <!-- This library can't ship stable until OpenAI ships stable. -->
+    <SuppressFinalPackageVersion>true</SuppressFinalPackageVersion>
+    <PackageTags>ai openai</PackageTags>
+    <Description>A client for OpenAI that integrates with Aspire, including metrics, logging and telemetry.</Description>
+    <PackageIconFullPath>$(SharedDir)OpenAI_256x.png</PackageIconFullPath>
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj:10]: Do we have/need permission to use this icon?

File: src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj [hunk 4]
```diff
@@ -0,0 +1,30 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(DefaultTargetFramework)</TargetFramework>
+    <IsPackable>true</IsPackable>
+    <!-- This library can't ship stable until OpenAI ships stable. -->
+    <SuppressFinalPackageVersion>true</SuppressFinalPackageVersion>
+    <PackageTags>ai openai</PackageTags>
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj:8]:
```suggestion
    <PackageTags>$(ComponentCommonPackageTags) ai openai</PackageTags>
```

File: src/Components/Aspire.OpenAI/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.OpenAI;
+using Aspire;
+using Microsoft.Extensions.Hosting;
+
+[assembly: ConfigurationSchema("Aspire:OpenAI", typeof(OpenAISettings))]
+[assembly: ConfigurationSchema("Aspire:OpenAI:ClientOptions", typeof(AspireOpenAIExtensions))]
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/AssemblyInfo.cs:9]:
```suggestion
[assembly: ConfigurationSchema("Aspire:OpenAI:ClientOptions", typeof(OpenAIClientOptions ))]
```

File: src/Components/Aspire.OpenAI/README.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+### Use configuration providers
+
+The .NET Aspire OpenAI library supports [Microsoft.Extensions.Configuration](https://learn.microsoft.com/dotnet/api/microsoft.extensions.configuration). It loads the `OpenAISettings` and `OpenAIClientOptions` from configuration by using the `Aspire:OpenAI` key. Example `appsettings.json` that configures some of the options:
+
+```json
+{
+  "Aspire": {
+    "OpenAI": {
+      "DisableTracing": false,
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/README.md:77]: We should also show how to configure the ClientOptions.

File: src/Components/Aspire.OpenAI/AspireOpenAIExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            settings.ParseConnectionString(connectionString);
+        }
+
+        configureSettings?.Invoke(settings);
+
+        var options = settings.Endpoint is null
+            ? new OpenAIClientOptions()
+            : new OpenAIClientOptions { Endpoint = settings.Endpoint };
+
+        configureOptions?.Invoke(options);
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/AspireOpenAIExtensions.cs:83]: The `options` should be bound to config as well.

File: src/Components/Aspire.OpenAI/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+[assembly: LoggingCategories()]
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/AssemblyInfo.cs:11]: Is this necessary? Or does it throw if it isn't specfiied?

File: tests/Aspire.OpenAI.Tests/ConformanceTests.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    protected override string ValidJsonConfig => """
+        {
+          "Aspire": {
+            "OpenAI": {
+                "Endpoint": "http://YOUR_URI",
+                "DisableTracing": false,
+                "DisableMetrics": false,
+                "ClientOptions": {
+                "NetworkTimeout": "00:00:02"
```
Comment (eerhardt) [inline tests/Aspire.OpenAI.Tests/ConformanceTests.cs:35]:
```suggestion
                    "NetworkTimeout": "00:00:02"
```

File: src/Components/Telemetry.md (hunk 2 above)
Comment (annelo-msft) [reply src/Components/Telemetry.md:261]: @m-redding is driving this work for SCM, we don't have a confirmed ship date for it yet, but I don't expect it to be a lot longer from here.

File: src/Components/Aspire.OpenAI/OpenAISettings.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Data.Common;
+
+namespace Aspire.OpenAI;
+
+/// <summary>
+/// The settings relevant to accessing OpenAI.
+/// </summary>
+public sealed class OpenAISettings
```
Comment (annelo-msft) [inline src/Components/Aspire.OpenAI/OpenAISettings.cs:11]: @KrzysztofCwalina, any concerns about adding this as a public type in this namespace?  I know you and I had discussed potentially putting such types in the SCM-based client namespaces themselves.

Or, @eerhardt, could you help us understand established precedents around these settings types for Aspire more broadly?  If we e.g. wanted to enable injection of an OpenAI client in a DI container outside the context of Aspire, would we want this type to live in a non-Aspire namespace?

File: src/Components/Aspire.OpenAI/README.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
+# Aspire.OpenAI library
+
+Registers [OpenAIClient](https://github.com/openai/openai-dotnet?tab=readme-ov-file#using-the-openaiclient-class) as a singleton in the DI container for using the OpenAI REST API. Enables corresponding metrics, logging and telemetry.
+
+## Getting started
+
+### Prerequisites
+
+- An OpenAI REST API compatible service like OpenAI.com, Azure AI OpenAI, ollama.com and others.
```
[the remark this answers is not in the document]
Comment (annelo-msft) [reply src/Components/Aspire.OpenAI/README.md:9]: There are differences between the REST APIs of the OpenAI and Azure OpenAI services that require using the different .NET client libraries to communicate with them.

File: src/Components/Aspire.OpenAI/README.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+```csharp
+private readonly OpenAIClient _client;
+
+public ChatController(OpenAIClient client)
+{
+    _client = client;
+}
+```
+
+To learn how to use the `OpenAIClient` class refer to [Using the OpenAIClient class](https://github.com/openai/openai-dotnet?tab=readme-ov-file#using-the-openaiclient-class).
```
Comment (annelo-msft) [inline src/Components/Aspire.OpenAI/README.md:39]: In Azure SDK terminology, we don't typically refer to clients as classes -- it is not technically incorrect to do so, but this sounds strange to my ears.  It might make sense to reword this to something like "learn how to use the `OpenAI` client library" ?

@AngelosP, can you recommend an official document the Aspire documents could link to for helping Aspire users onboard to the client?  @scottaddie, FYI that this may be an opportunity to build out docs about SCM-based clients that the Aspire team could leverage going forward.

File: src/Components/Aspire.OpenAI/README.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+```
+
+#### Connection string
+
+Alternatively, a custom connection string can be used.
+
+```json
+{
+  "ConnectionStrings": {
+    "openaiConnectionName": "Endpoint=https://{openai_rest_api_url}Key={account_key};"
```
[the remark this answers is not in the document]
Comment (annelo-msft) [reply src/Components/Aspire.OpenAI/README.md:60]: Does this work for the OpenAI client?  I was not aware that it supported connection strings, but if it does, I will update my mental model.

File: src/Components/Aspire.OpenAI/README.md (hunk 6 above)
Comment (annelo-msft) [reply src/Components/Aspire.OpenAI/README.md:77]: If you develop a general-purpose solution for SCM-based clients (which use `ClientPipelineOptions` instead of the Azure.Core `ClientOptions` type), let me know!  We are interested in this piece on our team.

@m-redding, @KrzysztofCwalina  FYI

File: src/Components/Aspire.OpenAI/README.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+You can also setup the `OpenAIClientOptions` using the optional `Action<OpenAIClientOptions>? configureOptions` parameter of the `AddOpenAIClient` method. For example, to set a custom `NetworkTimeout` value for this client:
+
+```csharp
+builder.AddOpenAIClient("openaiConnectionName", configureOptions: options => options.NetworkTimeout = TimeSpan.FromSeconds(2));
+```
+
+## AppHost extensions
+
+There is no specific AppHost extension corresponding to the OpenAI integration. Instead a connection string resource can be registered:
```
Comment (annelo-msft) [inline src/Components/Aspire.OpenAI/README.md:99]: For my learning -- what is an AppHost extension and is this something we would want to have a story for in the .NET ecosystem to support better ease of use for OpenAI?  Mostly just curious on this piece.

File: src/Components/Aspire.OpenAI/OpenAISettings.cs (hunk 10 above)
Comment (sebastienros) [reply src/Components/Aspire.OpenAI/OpenAISettings.cs:11]: These settings types are used in Aspire to drive what and how services are registered in DI. This is standard in Aspire, even components have it.

File: src/Components/Aspire.OpenAI/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (sebastienros) [reply src/Components/Aspire.OpenAI/README.md:39]: Just found a new page in docs, updating.

File: src/Components/Aspire.OpenAI/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (sebastienros) [reply src/Components/Aspire.OpenAI/README.md:60]: @annelo-msft this is specific to Aspire, there is no better way right now to pass settings across app boundaries than connection strings. So Aspire components sometimed make their own.

File: src/Components/Aspire.OpenAI/README.md (hunk 14 above)
Comment (sebastienros) [reply src/Components/Aspire.OpenAI/README.md:99]: An AppHost extension is a package that helps provision or describe the server part of the integration. For instance with Azure.AI.OpenAI there is a server resource that can be registered, will AI models too, such that we can provision the resource when the app is deployed. With OpenAI there is no such thing we can deploy, so we only declare a connection string, like a logical representation of this resource. Ultimately we could have a component that provides Ollama services for example.

File: src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj (hunk 3 above)
Comment (sebastienros) [reply src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj:10]: This is the file from the open GitHub repos and package but the repos is (c) OpenAI, so we can't use it as if Microsoft was the owner of the repos.

https://openai.com/brand/

> Use of the OpenAI logo & logo mark is limited to authorized uses only and the following conditions: ...

Which means we need an authorization. Do you suggest I use the Aspire logo instead?

File: src/Components/Aspire.OpenAI/AssemblyInfo.cs (hunk 8 above)
Comment (sebastienros) [reply src/Components/Aspire.OpenAI/AssemblyInfo.cs]: Removed. Works without it. I just wanted a placeholder to remember there are no logging categories.

File: src/Components/Aspire.OpenAI/AspireOpenAIExtensions.cs (hunk 7 above)
Comment (sebastienros) [reply src/Components/Aspire.OpenAI/AspireOpenAIExtensions.cs:87]: Options are now bound, test added.

File: src/Components/Aspire.OpenAI/OpenAISettings.cs (hunk 10 above)
Comment (annelo-msft) [reply src/Components/Aspire.OpenAI/OpenAISettings.cs:11]: If we wanted to also add a type called `OpenAISettings` in the `OpenAI` client namespace, would that be frowned upon?

File: src/Components/Aspire.OpenAI/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (annelo-msft) [reply src/Components/Aspire.OpenAI/README.md]: I believe @AngelosP would be the best person to share what are considered official docs for this library.

File: src/Components/Aspire.OpenAI/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (annelo-msft) [reply src/Components/Aspire.OpenAI/README.md]: Got it.  Thank you for the explanation!

File: src/Components/Aspire.OpenAI/OpenAISettings.cs (hunk 10 above)
Comment (eerhardt) [reply src/Components/Aspire.OpenAI/OpenAISettings.cs:11]: It wouldn't be great, but it would be possible. The namespace can disambiguate.

For Aspire, the pattern is `{NameOfLibrary}Settings` in the `Aspire.{NameOfLibrary}` namespace.

See

https://github.com/dotnet/aspire/blob/0490bcbc4318c22774d162b3f49382f148cfa6b2/src/Components/Aspire.RabbitMQ.Client/RabbitMQClientSettings.cs#L6-L9

https://github.com/dotnet/aspire/blob/0490bcbc4318c22774d162b3f49382f148cfa6b2/src/Components/Aspire.StackExchange.Redis/StackExchangeRedisSettings.cs#L6-L9

File: src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj (hunk 3 above)
Comment (eerhardt) [reply src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj:10]: @DamianEdwards - you've handled this for other integrations. Can you help here?

File: src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj [hunk 15]
```diff
@@ -0,0 +1,30 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(DefaultTargetFramework)</TargetFramework>
+    <IsPackable>true</IsPackable>
+    <!-- This library can't ship stable until OpenAI ships stable. -->
+    <SuppressFinalPackageVersion>true</SuppressFinalPackageVersion>
+    <PackageTags>$(ComponentCommonPackageTags) ai openai</PackageTags>
+    <Description>A client for OpenAI that integrates with Aspire, including metrics, logging and telemetry.</Description>
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj:9]:
```suggestion
    <Description>A client for OpenAI that integrates with Aspire, including metrics and telemetry.</Description>
```

No logging (yet).

File: src/Components/Aspire.OpenAI/AspireOpenAIExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+        if (builder.Configuration.GetConnectionString(connectionName) is string connectionString)
+        {
+            settings.ParseConnectionString(connectionString);
+        }
+
+        configureSettings?.Invoke(settings);
+
+        var options = new OpenAIClientOptions();
+        builder.Configuration.GetSection($"{configurationSectionName}:ClientOptions").Bind(options);
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/AspireOpenAIExtensions.cs:80]: I think we should use the "IOptions" pattern here, similar to how we do it in Redis:

https://github.com/dotnet/aspire/blob/0490bcbc4318c22774d162b3f49382f148cfa6b2/src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs#L89-L96

We are probably inconsistent in the other client libraries, but Redis and the Azure ones follow IOptions pattern.

File: src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj (hunk 3 above)
Comment (DamianEdwards) [reply src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj:10]: I'll defer to @maddymontaquila now :smile:

File: src/Components/Aspire.OpenAI/OpenAISettings.cs (hunk 10 above)
Comment (annelo-msft) [reply src/Components/Aspire.OpenAI/OpenAISettings.cs:11]: Cool, as long as we can keep that an an option in our back pocket, I have no concerns.

File: src/Components/Aspire.OpenAI/README.md (hunk 14 above)
Comment (annelo-msft) [reply src/Components/Aspire.OpenAI/README.md:98]: Thanks!  My comment here doesn't block this PR in any way, but I'm interested in understanding why we have one for Azure.AI.OpenAI and not OpenAI?  If there are pieces needed uniquely for SCM-based clients I'd be interested to understand the gaps there and if pieces are needed from SCM to address those gaps?

Comment (eerhardt) [reply src/Components/Aspire.OpenAI/README.md:98]: > why we have one for Azure.AI.OpenAI and not OpenAI?

Because you can provision an Azure OpenAI service (using bicep). That's what the "AppHost" extension will do.

You can't provision an OpenAI service. You need to sign up on their website.

File: src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj (hunk 3 above)
Comment (sebastienros) [reply src/Components/Aspire.OpenAI/Aspire.OpenAI.csproj]: Removing the logo for now until we have an answer.

**Comment by sebastienros** [comment]
Failing on unrelated cosmos tests

**Review by eerhardt — APPROVED** [review]
LGTM.

File: tests/Aspire.OpenAI.Tests/AspireOpenAIExtensionsTests.cs [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+        {
+            builder.AddOpenAIClient("openai");
+        }
+
+        using var host = builder.Build();
+        var client = useKeyed ?
+            host.Services.GetRequiredKeyedService<OpenAIClient>("openai") :
+            host.Services.GetRequiredService<OpenAIClient>();
+
+        Assert.NotNull(client);
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply tests/Aspire.OpenAI.Tests/AspireOpenAIExtensionsTests.cs:41]: I meant, can we assert something like the connection string or URI is correct?

[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/Aspire.OpenAI.Tests/AspireOpenAIExtensionsTests.cs:41]: That was my initial thought too but I couldn't find it in the model, do you know if it's somewhere? I didn't dig super deep, maybe I missed it. Will check again.

[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/Aspire.OpenAI.Tests/AspireOpenAIExtensionsTests.cs:41]: Are you fine with private reflection in this case?

[the remark this answers is not in the document]
Comment (eerhardt) [reply tests/Aspire.OpenAI.Tests/AspireOpenAIExtensionsTests.cs:41]: Nah, let's just live with what we have. If we mis-configured it, we will know because things won't work.
